### PR TITLE
DAOS-9145 control: Fix race when accessing runner cmd field

### PR DESCRIPTION
**Pull request by tanabarr**
Race detected when accessing cmd object that represents the underlying
OS process. Fix by adding a mutex for the runner cmd field and locking
appropriately.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

**Review by kjacque — APPROVED** [review]

File: src/control/server/engine/exec.go [hunk 1]
```diff
@@ -74,14 +76,18 @@ func (r *Runner) run(ctx context.Context, args, env []string, errOut chan<- erro
...
 	r.cmd = cmd
+	r.cmdMut.Unlock()
 
 	waitDone := make(chan struct{})
 	go func() {
 		r.running.SetTrue()
 		defer r.running.SetFalse()
 
-		errOut <- errors.Wrapf(common.GetExitStatus(cmd.Wait()), "%s exited", binPath)
+		r.cmdMut.Lock()
```
Comment (kjacque) [inline src/control/server/engine/exec.go:88]: Actually, hm, should this be an RLock?

File: src/control/server/engine/exec.go [hunk 2]
```diff
@@ -128,12 +136,16 @@ func (r *Runner) Signal(signal os.Signal) error {
 
 	r.log.Debugf("Signalling I/O Engine instance %d (%s)", r.Config.Index, signal)
 
+	r.cmdMut.Lock()
```
Comment (kjacque) [inline src/control/server/engine/exec.go:139]: Same question as above.

**Review by mjmac — CHANGES_REQUESTED** [review]
All control plane PRs must have the `Features: control` pragma, so at a minimum that needs to be fixed.

Second of all, I am not all that keen on solving the problem here with yet another mutex. I think it's just inherently racy to be working with a struct member across multiple goroutines, and I suspect that we won't see the end of these problems with this approach.

As an alternative, I pushed a PR that uses channels to communicate with a single goroutine that's responsible for managing the `exec.Cmd` instance. On the one hand this may seem more complicated, but logically it's much easier to reason about access patterns, IMO. You are welcome to take it or leave it, but either way, I don't think that just adding a mutex is going to be a long-term solution. 

https://github.com/daos-stack/daos/pull/8208

**Comment by tanabarr** [comment]
@mjmac happy for you to go ahead with your PR, will close this one in preference of yours and will reassign the ticket to you, thanks.
this PR is replaced by https://github.com/daos-stack/daos/pull/8208

File: src/control/server/engine/exec.go (hunk 1 above)
Comment (tanabarr) [reply src/control/server/engine/exec.go:88]: Wait() modifies process state (writes)

File: src/control/server/engine/exec.go (hunk 2 above)
Comment (tanabarr) [reply src/control/server/engine/exec.go:139]: this one probably could be because the Signal() call does not write an internal state.

**Pull request closed without merging**